cli/internal/cmd/upgradeapply.go

```diff
@@ -104,37 +117,38 @@ func (u *upgradeApplyCmd) upgradeApply(cmd *cobra.Command, fileHandler file.Hand
 		cmd.PrintErrln("WARNING: Skipping service and image upgrades, which are currently only supported for Azure and GCP.")
 	}
 
-	// If an image upgrade was just executed there won't be a diff. The function will return nil in that case.
-	if err := u.upgradeMeasurementsIfDiff(cmd, conf.GetMeasurements(), flags); err != nil {
-		return fmt.Errorf("upgrading measurements: %w", err)
-	}
-
 	return nil
 }
 
-// upgradeMeasurementsIfDiff checks if the locally configured measurements are different from the cluster's measurements.
+// upgradeAttestConfigIfDiff checks if the locally configured measurements are different from the cluster's measurements.
 // If so the function will ask the user to confirm (if --yes is not set) and upgrade the measurements only.
-func (u *upgradeApplyCmd) upgradeMeasurementsIfDiff(cmd *cobra.Command, newMeasurements measurements.M, flags upgradeApplyFlags) error {
-	clusterMeasurements, _, err := u.upgrader.GetClusterMeasurements(cmd.Context())
+func (u *upgradeApplyCmd) upgradeAttestConfigIfDiff(cmd *cobra.Command, newConfig config.AttestationCfg, flags upgradeApplyFlags) error {
+	clusterAttestationConfig, _, err := u.upgrader.GetClusterAttestationConfig(cmd.Context(), newConfig.GetVariant())
 	if err != nil {
-		return fmt.Errorf("getting cluster measurements: %w", err)
-	}
-	if clusterMeasurements.EqualTo(newMeasurements) {
-		return nil
+		// Config migration from v2.7 to v2.8 requires us to skip comparing configs if the cluster is still using the legacy config.
+		// TODO: v2.9 Remove error type check
+		if !errors.Is(err, kubernetes.ErrLegacyJoinConfig) {
+			return fmt.Errorf("getting cluster measurements: %w", err)
+		}
+	} else {
+		// If the current config is equal, or there is an error when comparing the configs, we skip the upgrade.
+		if equal, err := newConfig.EqualTo(clusterAttestationConfig); err != nil || equal {
+			return err
+		}
```

Comment:
It took me a moment to understand the control flow. Maybe the following version is slightly easier to understand? Not sure..
```
clusterAttestationConfig, _, err := u.upgrader.GetClusterAttestationConfig(cmd.Context(), newConfig.GetVariant())
// TODO: v2.9 Remove error type check
// Config migration from v2.7 to v2.8 requires us to skip comparing configs if the cluster is still using the legacy config.
if err != nil && !errors.Is(err, kubernetes.ErrLegacyJoinConfig) {
	return fmt.Errorf("getting cluster measurements: %w", err)
}
// TODO: v2.9 remove check, keep comparison
if err == nil {
	// If the current config is equal, or there is an error when comparing the configs, we skip the upgrade.
	equal, err := newConfig.EqualTo(clusterAttestationConfig)
	if err != nil || equal {
		return err
	}
}
```